src/transport/p2p.cu

```diff
@@ -487,10 +482,62 @@ end:
   } while (0)
 
 /* Send: Create and return connect structures for this peer to connect to me */
+static ncclResult_t getGpuHdpReg(int cudaDev, uint32_t** hdp) {
+  auto convert_bdf = [](const char *busId) {
+    char bdf[9];
+    strncpy(bdf, busId, 4);
+    strncpy(bdf+4, busId+5, 2);
+    strncpy(bdf+6, busId+8, 2);
+    bdf[8] = '\0';
+    uint16_t id = (uint16_t)strtol(bdf, NULL, 16);
+    return id;
+  };
+
+  const auto& find_agent = [](hsa_agent_t agent, void* out) {
+    hsa_agent_t* found_agent = (hsa_agent_t*)out;
+    uint16_t id = (uint16_t)found_agent->handle;
+    hsa_device_type_t type;
+    hsa_agent_get_info(agent, HSA_AGENT_INFO_DEVICE, (void*)&type);
+    if(type == HSA_DEVICE_TYPE_GPU) {
+      uint16_t bdf_id = 1;
+      hsa_agent_get_info(agent, (hsa_agent_info_t)HSA_AMD_AGENT_INFO_BDFID, &bdf_id);
+      if(bdf_id == id) {
+        *found_agent=agent;
+        return HSA_STATUS_ERROR;
+      }
+    }
+    return HSA_STATUS_SUCCESS;
+  };
+
+  char busId[NVML_DEVICE_PCI_BUS_ID_BUFFER_SIZE];
+  *hdp = NULL;
+  CUDACHECK(hipDeviceGetPCIBusId(busId, NVML_DEVICE_PCI_BUS_ID_BUFFER_SIZE, cudaDev));
+  hsa_agent_t agent;
+  agent.handle = convert_bdf(busId);
+  hsa_iterate_agents(find_agent, (void*)&agent);
+  hsa_amd_hdp_flush_t hdpinfo;
+  hsa_status_t err = hsa_agent_get_info(agent, (hsa_agent_info_t)HSA_AMD_AGENT_INFO_HDP_FLUSH, &hdpinfo);
```

Comment:
If agent lookup on 517 fails to locate the device then this call will almost certainly segfault since agent.handle will have non-null garbage in it.  Adding a check (or at least assert) will help debugging if something does go wrong in bdf id comparison.

Reply:
done